sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala

```diff
@@ -248,11 +250,17 @@ class HadoopTableReader(
         // SPARK-13709: For SerDes like AvroSerDe, some essential information (e.g. Avro schema
         // information) may be defined in table properties. Here we should merge table properties
         // and partition properties before initializing the deserializer. Note that partition
-        // properties take a higher priority here. For example, a partition may have a different
-        // SerDe as the one defined in table properties.
+        // properties take a higher priority here except for the Avro table properties when
+        // "spark.sql.hive.avroSchemaEvolution.enabled" is set because in that case the properties
+        // given at table level will be used (for details please check SPARK-26836).
+        // For example, a partition may have a different SerDe as the one defined in table
+        // properties.
         val props = new Properties(tableProperties)
-        partProps.asScala.foreach {
-          case (key, value) => props.setProperty(key, value)
+        partProps.asScala.foreach { case (key, value) =>
+          if (!avroSchemaEvolutionEnabled ||
+            !avroTablePropertyKeys.contains(key) || !tableProperties.containsKey(key)) {
```

Comment:
Can we rephrase this as `!(avroSchemaEvolutionEnabled && avroTablePropertyKeys.contains(key) && tableProperties.containsKey(key))` ? I think the disjunction of negations is more confusing.

You can also consider something like:
```
partProps.asScala
  .filterNot { case (key, _) => avroSchemaEvolutionEnabled && avroTablePropertyKeys.contains(key) && ... }
  .foreach { case (key, value) => props.setProperty(key, value)
```
but I'm not sure if that is more readable or not.

Reply:
I like the more functional solution with the `filterNot`.